staticd: BFD static route monitoring

rzalamena opened the pull request:
This PR is a refresh of PRs #7101 , #9704 and #10230 .

---

Add `staticd` integration with BFD to support dynamic install/remove of static routes based on BFD session monitoring. The following new `ip route` and `ipv6 route` command variations were added:
- `ip route A.B.C.D/M A.B.C.D bfd [{multi-hop|source A.B.C.D|profile PROFILE_NAME}]`
- `ipv6 route A.B.C.D/M A.B.C.D bfd [{multi-hop|source A.B.C.D|profile PROFILE_NAME}]`

The monitored static routes will only be installed to RIB if the BFD session it is linked to is in `up` state.

Changes since previous PRs:
- Removed the `route group` code: this is not needed for basic usage and can be added later in a separated PR if wanted.
- Fixed some crashes and issues found during testing
- Rewrote the BFD automatic source selection: instead of implementing it per-daemon it is now part of the FRR library (see `lib/bfd.c`)
- Reworded the `doc` and removed `route group` from `doc`
- Excluded the "auto hop" commit from 6Wind (I don't think it is necessary, just use profile to select minimum TTL)
- Simplified the YANG model: all BFD options under a single grouping.

Thanks to @pguibert6WIND and @louis-6wind for the previous fixes and improvements.

rzalamena commented:
frrbot is complaining about `DEFPY` formatting which I won't change to avoid breaking the style with every other `DEFUN`s in the code base.

louis-6wind commented:
> * Excluded the "auto hop" commit from 6Wind (I don't think it is necessary, just use profile to select minimum TTL)

Thank you for the work continuation. Auto-hop selects the type of BFD automatically: single-hop or multi-hop. It is not about TTL

rzalamena commented:
Rebased to pull the latest topotest fixes and added a new commit to help `lib/bfd.c` clean up on shutdown (see commit `lib: BFD integration clean up function`).

riw777 reviewed (approved):
looks good

louis-6wind commented on yang/frr-bfdd.yang:
```diff
@@ -242,6 +242,40 @@ module frr-bfdd {
+    description
+      "BFD monitoring template for protocol integration.";
+
+    leaf enable {
```
it seems that when bfd-monitoring only contains "enable true", it is single-hop. It is not clear enough.

`source`, `multi-hop` and `profile` cannot be set together. You need have `choice`

```
choice type {
  case single-hop {
    leaf single-hop {
      type boolean;
    }
    leaf source {
      (...)
    }
  }
  case multi-hop {
    leaf multi-hop {
      type boolean;
    }
  }
  case profile {
    leaf profile {
      description
        "BFD pre configured profile.";
      type frr-bfdd:profile-ref;
    }
  }
```
In that way, it will be easy to add the `auto-hop` option to automatically select single-hop or multi-hop

rzalamena replied:
`enable` only toggles BFD session status (`true` or `false`). The `multi-hop` toggle is a leaf one item below (defaults to `false`), you can set it to `true` if you want.

You can always use `profile` if you want.

I'm very confused by this `choice type` you are proposing.

louis-6wind replied:
> You can always use `profile` if you want.

OK. You can also have a `source` with `multi-hop`. No need for `choice` then.

However, `enable` is confusing. It is the administrative state of the BFD session but there is no CLI option for that. You never do `enable false` but you remove the BFD monitored route in the CLI.

I think the following is a better option.

```
        leaf hop-mode {
          description
            "Hop mode";
          default "single-hop";
          type enumeration {
              enum single-hop {
              value 0;
              description
                "single-hop BFD session";
            }
            enum multi-hop {
            value 1;
              description
                "multi-hop BFD session";
            }
          }
        }
```

In the future:

```
            enum auto {
              value 2;
              description
                "Select automatically the BFD mode: single or multi-hop";
            }
```

rzalamena replied:
> However, enable is confusing. It is the administrative state of the BFD session but there is no CLI option for that. You never do enable false but you remove the BFD monitored route in the CLI.

Unfortunately the `enable` toggle doesn't exist for static route CLI, but it will for other daemons integration. For an example: `bgpd`. See discussion here: https://github.com/FRRouting/frr/pull/11951#issuecomment-1330776764

I'm implementing it here because it will eventually be used and the northbound handling is also implemented for `staticd` (even though there is no CLI).

> I think the following is a better option.
> ```yang
> leaf hop-mode {
> ```

Sorry I'm not implementing auto hop, but I can help you implement/review it later if you want.

louis-6wind replied:
I did not find where you set enable in bgp CLI:
You do not do `neighbor 172.16.1.1 bfd enabled` but just `neighbor 172.16.1.1 bfd`. If you want to disable BFD, you just remove the command.

rzalamena replied:
@louis-6wind it doesn't exist yet. The new `grouping bfd-monitoring` will be a template that protocols should use to integrate with `bfdd`. In BGP's case `neighbor A.B.C.D bfd` will toggle the `enable` leaf, today it doesn't even use YANG/northbound.

Same will apply to `ospfd`, `ospf6d`, `pimd`, `isisd` and `fabricd`:
- `ip ospf bfd`
- `ipv6 ospf6 bfd`
- `isis bfd`
- `ip pim bfd`

This is needed to satisfy the condition ( https://github.com/FRRouting/frr/pull/11951#discussion_r975526265 ) of allowing protocols to save BFD configuration but not have it enabled. After these changes these protocols should be able to configure `ip ospf bfd profile FOO`, for an example, and save configuration without having BFD enabled.

rzalamena replied:
> enable is not needed unless you have a CLI enable option, which is not the case.

Sure. I'll remove it because it doesn't make sense for this particular case, but when the other daemons work start it will come back anyway.

I'll push the changes soon.

> May be moving container bfd-monitoring under grouping bfd-monitoring would be better

Makes sense, I didn't think of this before.

rzalamena replied:
@louis-6wind I removed the `enable` leaf as requested.

I could not move the `bfd-monitoring` container to the grouping and it didn't make sense because otherwise we'd introduce an extra container level. The outer container in `frr-staticd.yang` is required because I introduced some conditionals to allow the existence of BFD: BFD monitoring can only be used with IPv4/IPv6 gateways / routes. I think other protocols might want to add conditionals as well, so let the user protocol decide how the YANG tree looks like.

louis-6wind replied:
OK. Thank you for the change

rzalamena commented on staticd/static_vty.c:
```diff
@@ -1165,6 +1317,26 @@ static void nexthop_cli_show(struct vty *vty, const struct lyd_node *route,
 		vty_out(vty, " color %s",
 			yang_dnode_get_string(nexthop, "./srte-color"));
 
```
It's not missing, it is at the very end of this function. Right after the closing bracket of `if (yang_dnode_exists(nexthop, "./bfd-monitoring") && yang_dnode_get_bool(nexthop, "./bfd-monitoring/enable")) {`.

louis-6wind replied:
OK

ton31337 commented:
ci:rerun problems with SNMP-PDU on mininet2503.lab.netdef.org

rzalamena commented:
CI:rerun unrelated ospf_asbr_summary_topo1 failure

rzalamena commented:
CI:rerun unrelated ospfv3_rte_calc failure

louis-6wind commented on yang/frr-bfdd.yang:
```diff
@@ -242,6 +242,33 @@ module frr-bfdd {
+
+         When source is not set and and multi-hop is `true` the source
+         address will be automatic selected through Next Hop Tracking (NHT).";
+    }
```
Please add a note about single-hop as well

rzalamena replied:
I added note about single hop.

louis-6wind replied:
fine

riw777 reviewed (approved):
looks good ... waiting on other reviewers

riw777 commented:
ci failures in convergence ... I don't see how they're related, so I'll rerun

riw777 commented:
lots of ci failures on table versions, which shouldn't have anything to do with this pr (?) ... trying again

riw777 commented:
ci:rerun

mwinter-osr commented:
CI:rerun

louis-6wind commented:
@rzalamena I cannot apply two static routes with the same nexthop

```
ipv6 route 2001:db8:5::/64 2001:db8:4::3 bfd multi-hop profile slow-tx
ipv6 route 2001:db8:6::/64 2001:db8:4::3 bfd multi-hop profile slow-tx
```
Second route is not installed
```
r3# show bfd static route
Showing BFD monitored static routes:

  Next hops:
    VRF default IPv4 Unicast:

    VRF default IPv4 Multicast:

    VRF default IPv6 Unicast:
        2001:db8:5::/64 peer 2001:db8:4::3 (status: installed)
        2001:db8:6::/64 peer 2001:db8:4::3 (status: uninstalled)
r3(config)# do sh ipv6 route
Codes: K - kernel route, C - connected, S - static, R - RIPng,
       O - OSPFv3, I - IS-IS, B - BGP, N - NHRP, T - Table,
       v - VNC, V - VNC-Direct, A - Babel, F - PBR,
       f - OpenFabric,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
       t - trapped, o - offload failure

B>* 2001:db8:1::/64 [20/0] via 2001:db8:1::1 inactive, weight 1, 00:00:21
  *                        via fe80::482d:d1ff:fe87:83d4, r3-eth0, weight 1, 00:00:21
C>* 2001:db8:2::/64 is directly connected, r3-eth0, 00:00:24
C>* 2001:db8:3::/64 is directly connected, r3-eth1, 00:00:25
B>* 2001:db8:4::/64 [20/0] via fe80::801f:13ff:fead:f552, r3-eth1, weight 1, 00:00:20
S>  2001:db8:5::/64 [1/0] via 2001:db8:4::3 (recursive), weight 1, 00:00:19
  *                         via fe80::801f:13ff:fead:f552, r3-eth1, weight 1, 00:00:19
C * fe80::/64 is directly connected, r3-eth0, 00:00:24
C * fe80::/64 is directly connected, r3-eth1, 00:00:25
C>* fe80::/64 is directly connected, lo, 00:00:26
r3(config)# exit
r3# show s
segment-routing  sharp            sr-te            startup-config
r3# show bfd peers brief
Session count: 4
SessionId  LocalAddress                             PeerAddress                             Status
=========  ============                             ===========                             ======
2377889738 2001:db8:2::1                            2001:db8:2::2                           up
3803863359 2001:db8:3::2                            2001:db8:3::1                           up
534538474  192.168.2.1                              192.168.1.1                             up
2643518653 2001:db8:3::2                            2001:db8:4::3                           up

```
